### PR TITLE
Upgrade to support GitHub's new "Projects V2" status board

### PR DESCRIPTION
**Pull request by rbuckton**
This makes a number of changes to support GitHub's "Projects V2" project boards, which can only be accessed via the GitHub graphql API. Since I can't filter "Projects V2" results like I could "classic" projects, I've had to make some changes to the underlying query mechanism to fetch *all* project items and filter them client side, as the GitHub graphql search API does not yet seem to support "Projects V2"-specific search criteria.

### CODE REVIEW

File: src/github.graphql
Comment (jakebailey) [inline src/github.graphql:1]: Is there a reason to commit this file rather than just pulling it as a dep?

File: package.json [hunk 1]
```diff
@@ -21,16 +21,20 @@
     "url": "https://github.com/DefinitelyTyped/focus-dt/issues"
   },
   "devDependencies": {
+    "@octokit/rest": "^21.0.0",
```
Comment (jakebailey) [inline package.json:24]: You shouldn't need `@octokit/rest` if you're using `octokit` itself.

File: package.json [hunk 2]
```diff
@@ -21,16 +21,20 @@
...
     "@types/color-convert": "^2.0.0",
     "@types/node": "^16.11.47",
     "@types/winreg": "^1.2.30",
     "@types/yargs": "^13.0.0",
-    "typescript": "^4.7.4"
+    "octokit": "^4.0.2",
```
Comment (jakebailey) [inline package.json:29]: Was this intended to be a prod dep?

File: package.json [hunk 3]
```diff
@@ -21,16 +21,20 @@
...
+    "typescript": "^5.5.3"
   },
   "dependencies": {
-    "@octokit/rest": "^18.5.2",
+    "@0no-co/graphqlsp": "^1.12.11",
+    "@octokit/graphql-schema": "^15.25.0",
```
Comment (jakebailey) [inline package.json:34]: I think these are both dev deps?

File: src/github.graphql
Comment (rbuckton) [reply src/github.graphql:1]: Yeah, I made some minor changes to the schema where I'm pretty sure the original schema authors forgot a `!`, such as in `ProjectV2ItemConnection`:

```graphql
type ProjectV2ItemConnection {
  ...
  nodes: [ProjectV2Item] # theirs
  nodes: [ProjectV2Item!]! # mine
  ...
}
```

I expect it's unlikely that the `nodes` array will be `null` or have `null` entries.

File: package.json (hunk 2 above)
Comment (rbuckton) [reply package.json]: Fixed, thanks. I was too aggressive with `-D`

File: package-lock.json [hunk 4]
```diff
@@ -9,9 +9,11 @@
       "version": "1.0.4",
       "license": "Apache-2.0",
       "dependencies": {
-        "@octokit/rest": "^18.5.2",
+        "@0no-co/graphqlsp": "^1.12.11",
+        "@octokit/graphql-schema": "^15.25.0",
         "chalk": "^3.0.0",
         "equatable": "^1.2.0",
+        "gql.tada": "^1.8.2",
```
Comment (jakebailey) [inline package-lock.json:16]: The lockfile seems to be out of sync with package.json (could use another npm install)

File: package.json [hunk 5]
```diff
@@ -20,18 +20,23 @@
   "bugs": {
     "url": "https://github.com/DefinitelyTyped/focus-dt/issues"
   },
+  "engines": {
+    "node": "^20.0.0"
```
Comment (jakebailey) [inline package.json:24]:
```suggestion
    "node": ">=20.0.0"
```

(or `>=20`)

I use Node 22 so this would avoid some sort of warning.

**Comment by jakebailey** [comment]
Hm, tried this locally and it looks it thinks I'm missing a scope but I do, it's just that I have the write scope too?

```
Your authentication token is not authorized for the requisite scope 'read:project'. Please replace the token and try again.
PS D:\work\focus-dt> gh auth status
github.com
  ✓ Logged in to github.com account jakebailey (keyring)
  - Active account: true
  - Git operations protocol: ssh
  - Token: gho_************************************
  - Token scopes: 'gist', 'project', 'read:org', 'repo'
```

File: src/init.ts [hunk 6]
```diff
@@ -36,18 +52,29 @@ export async function init() {
...
+    if (!status.authenticated) {
+        checkGhInstalled();
+        token = ghAuthLogin();
+        status = await ghAuthStatus(token);
+        checkAuthenticated(status);
+    }
+    if (userSuppliedToken) {
+        checkRequiredScopes(status);
+    }
+    else if (!status.scopes?.includes("read:project")) {
```
Comment (jakebailey) [inline src/init.ts:68]: I think this is the same as `checkRequiredScopes` below.

Comment (jakebailey) [reply src/init.ts:68]: Ah, same check, but that one throws.

Comment (rbuckton) [reply src/init.ts]: I guess it can't be as easy the header returning the full set of scopes.

**Comment by rbuckton** [comment]
A few findings from this endeavor:

- The GitHub graphql schema seems to lack `!` (non-null) markers in a few places where it doesn't seem like a `null` value is likely. Using the schema with a type generator like `gql.tada` results in types that are fairly difficult to work with due to the extreme number of potentially `null` types.

- The API for "Projects V2" is woefully under-documented and there are no clear examples in the developer documentation. In the classic projects REST API, it was clear how to query "columns" and "cards". In the "Projects V2" UI, columns are defined by a "Column by" option in the view:
image
However, in the API this is referred to as a "vertical group-by field".

- There does not seem to be a way to filter the project items in a project. I thought that maybe the `search` API might be used for this, but the `search` API does not seem to support "Projects V2" criteria yet.

- It took far too long to figure out that a `GitObjectID` in the API is equivalent to the SHA commit hash used by the REST API as it isn't clearly labeled as such in the developer documentation or the graphql schema.

- For all of its power, `gql.tada` is extremely computationally expensive and results in very slow response times for completions in the editor.

**Comment by jakebailey** [comment]
> * For all of its power, `gql.tada` is extremely computationally expensive and results in very slow response times for completions in the editor.

Definitely agree, through they do have "turbo mode" which helps when the queries are changing.

File: src/init.ts [hunk 7]
```diff
@@ -82,9 +109,42 @@ export async function init() {
...
+}
+
+function checkAuthenticated(status: AuthStatus) {
+    if (!status.authenticated) {
+        failNotAuthenticated();
+    }
+}
+
+function hasRequiredScopes(status: AuthStatus) {
+    return !!status.scopes?.includes("read:project") || !status.scopes?.includes("project");
```
Comment (jakebailey) [inline src/init.ts:143]:
```suggestion
    return !!status.scopes?.includes("read:project") || !!status.scopes?.includes("project");

```

**Review by jakebailey — APPROVED** [review]

**Pull request merged**